### ISSUE TITLE
Can't use intermediate layer as input for a NN within a NN?

### ISSUE THREAD

**Issue by dc-ai** [fchollet/keras#5329]
I'm trying to create a neural network "within" a neural network, basically I branch off from an intermediate layer, use that as an output to train another neural network. I want to train both networks together. I can't seem to do it, I run into the following error:

_UserWarning: Model inputs must come from a Keras Input layer, they cannot be the output of a previous non-Input layer_

Sample code below:

```
# 1st NN: Simple convolutional autoencoder
input_img = Input(shape=(1, 80, 80), name="input_layer")
intermediate = Convolution2D(32, 3, 3, subsample=(2, 2))(input_img)
decoded = Deconvolution2D(1, 3, 3, subsample=(2, 2))(intermediate)

#2nd NN: Uses output from encoding part to do supervised learning on FC layers
fc_flatten = Flatten()(intermediate)
y = Dense(64)(fc_flatten)
fc_output = Dense(1)(y)

model = Model(input=[input_img, fc_flatten], output=[decoded, fc_output])
```

I don't understand. I'm using almost the same architecture as the "Multi-input and multi-output models" example on the Keras Functional API, with the exception that I don't have an auxillary input.

**Comment by siddheshk** [comment]
It seems like in the line
`model = Model(input=[input_img, fc_flatten], output=[decoded, fc_output])`
You are passing `fc_flatten` as an input. As `fc_flatten` is dependent on `intermediate`, it should work if you remove it from the input parameter.
Something like 
`model = Model(input=[input_img], output=[decoded, fc_output])`

**Comment by dc-ai** [comment]
Thanks, siddheshk, but there is a reason why I set it up that way.

I am calculating loss functions used in training that compares input_img vs decoded, and fc_flatten vs fc_output.

If I remove fc_flatten from the input, I won't be able to calculate its loss function and use it for training.

**Comment by unrealwill** [comment]
Hello, 
You almost got it. Use this little trick : 
# 1st NN: Simple convolutional autoencoder
input_img = Input(shape=(1, 80, 80), name="input_layer")
intermediate = Convolution2D(32, 3, 3, subsample=(2, 2))(input_img)
decoded = Deconvolution2D(1, 3, 3, subsample=(2, 2))(intermediate)

fc_flatten = Flatten()(intermediate)

#2nd NN: Uses output from encoding part to do supervised learning on FC layers
input_NN2 = Input( your correct shape )
y = Dense(64)(input_NN2)
fc_output = Dense(1)(y)
model2 = Model( [input_NN2],[fc_output])

fc_output = model2( fc_flatten )
model = Model(input=[input_img, fc_flatten], output=[decoded, fc_output])

**Comment by dc-ai** [comment]
Thanks, unrealwill, that got me one stop closer. However on training the model, keras still gives the following error:
_ValueError: The model expects 2 input arrays, but only received one array_

It looks like Keras only sees the original image (that's the 1st input array). The 2nd input array is generated by model2 during the training process, and it looks like Keras needs it to be there at the start?

I'm a bit confused, any help is appreciated. Thanks.

**Comment by unrealwill** [comment]
Can you please check that you are indeed using 
model = Model(input=[input_img], output=[decoded, fc_output])
I edited my code to remove the second input which is unneeded.

**Comment by dc-ai** [comment]
Setting it up either with 1 or 2 inputs:
model = Model(input=[input_img], output=[decoded, fc_output])
or
model = Model(input=[input_img, fc_flatten], output=[decoded, fc_output])

Did not change the results, same error message. I am only feeding in the 80*80 image data (i.e. input_img). The second input, fc_flatten is being dynamically generated as the model gets trained.

Keras error messages seems to suggest that I have parse in both input_img and input_NN2 in the model.fit() command?

**Comment by unrealwill** [comment]
This code (Not exactly the code you want I use convolution instead of deconvolution and changed the shape to calculate more easily the shape) works for me : 

```
def buildInnerModel():
    # 1st NN: Simple convolutional autoencoder
    input_img = Input(shape=(1, 80, 80), name="input_layer")
    intermediate = Convolution2D(32, 3, 3, border_mode="same")(input_img)
    decoded = Convolution2D(10, 3, 3, border_mode="same")(intermediate)

    fc_flatten = Flatten()(intermediate)

    # 2nd NN: Uses output from encoding part to do supervised learning on FC layers
    input_NN2 = Input( (None,80*80*32))
    y = Dense(64)(input_NN2)
    fc_output = Dense(1)(y)
    model2 = Model([input_NN2], [fc_output])

    fc_output = model2([fc_flatten])
    model = Model(input=[input_img], output=[decoded, fc_output])
    model.compile( loss=["mse","mse"],optimizer="adam")
    return model


model = buildInnerModel()
print model.predict( np.random.randn(10,1,80,80) )[0].shape
#(10, 10, 80, 80)
print model.predict( np.random.randn(10,1,80,80) )[1].shape
#(10, 1)
model.fit( [np.random.randn(10,1,80,80)],[np.random.randn(10,10,80,80),np.random.randn(10,1)])

```

**Comment by dc-ai** [comment]
Thanks, that worked although it's not exactly what I need. I managed to find another workaround partially inspired by your code, but that itself is also causing problems further downstream.

Back to the issue, my lingering question is that I still can't find a way to use intermediate layer output as input in Keras, and train it at the same time.

What this means is that the architecture is strictly speaking, single-input to multiple-output. However, additional inputs are being generated by the intermediate layers and used to train another part of the neural network.

For the code example you provided
`model.fit( [np.random.randn(10,1,80,80)],[np.random.randn(10,10,80,80),np.random.randn(10,1)])`

It seems to indicate you have two sources of data, one of dimensions (1, 80, 80) and the other of dimension (1). I don't have that. I only have one data source of dimension (1, 80, 80).

**Comment by unrealwill** [comment]
You need to "explode" your model into two models to expose your intermediate layer.
inp = Inp(inputshape)
lay1=Layer1()(inp)
layer2 Layer2()(lay2)
intermModel=Model( inp, lay2 )

inp = Inp(intermediateshape)
layer3A = Layer3A()(inp)
layer3B = Layer3B()(layer3A
secondHalf1Model = Model( inp, layer3B)

inp = Inp(intermediateshape)
layer4A = Layer4A()(inp)
layer4B = Layer4B()(layer3A
secondHalf2Model = Model( inp, layer4B)

inp = Inp(inputshape)
outputA = secondHalf1Model( interModel( inp ) )
outputB = secondHalf2Model( interModel( inp ) )

combinedModel = Model( {inp], [outputA,outputB] )

**Comment by ckyleda** [comment]
This remains a problem in 2019 - why on earth can't we connect layers together exactly how we want, without these half-working workarounds?

Layer connections should not be dependent upon Keras' idea of what an 'input' is - at least as long as there's an input further up the chain!

**Comment by unrealwill** [comment]
@ckyleda : I guess it's a design decision : Model encapsulation. 
Which is very useful because it means you can substitute one model by another without having to worry about those inside layers.

What is indeed problematic is that this useful by model decomposition is lost during serialization, but you are free to generate your model with code and load your weights manually.

But if you are grasping weights from a model zoo, you have to recreate the model architecture with code to create the ops, then you fill the weights of the variable using the matching names. 

If you are using tensorflow directly you can have more control of what you want to do with variable_scope, though they are being deprecated in tensorflow 2.0. 

Keras can't do these kind of things while remaining theano compatibility as variable_scope don't exist there.

**Comment by ckyleda** [comment]
Once again I find myself with this issue.
In my opinion theano compatibility should be dropped, as theano is dead and long gone.

**Comment by sudonto** [comment]
Hi @unrealwill , I have similar question to the original question. What I want is to feed a model with an output from intermediate layer. Usually we feed `fit_generator()` with external data, but I would like to use tensor coming from intermediate layer. Is this possible?

I am using Keras 2.2.4 and TF 1.12.0

**Comment by iulialexandra** [comment]
I have the same question. Has anything changed in more recent versions of Keras? I'd rather avoid making a new env to explore this if possible.
